File: compat/suites/dotnet-sdk/Harness/Runner.cs
using System.Text.Json;

namespace OvercastCompat.Harness;

public static class Runner
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull,
    };

    public static async Task RunSuiteAsync(string suite, string endpoint, string region, IReadOnlyList<TestGroup> groups)
    {
        var started = DateTimeOffset.UtcNow;
        Emit(new
        {
            @event = "run_start",
            suite,
            started_at = started.ToString("O"),
            endpoint,
            version = "1",
            total_tests = groups.Sum(group => group.Tests.Count),
        });

        var slots = 8;
        if (int.TryParse(Environment.GetEnvironmentVariable("OVERCAST_COMPAT_PARALLEL_SLOTS"), out var configured) && configured > 0)
        {
            slots = configured;
        }

        var semaphore = new SemaphoreSlim(slots, slots);
        var tasks = groups.Select(async group =>
        {
            await semaphore.WaitAsync();
            try
            {
                return await RunGroupAsync(suite, endpoint, region, group);
            }
            finally
            {
                semaphore.Release();
            }
        });

        var results = await Task.WhenAll(tasks);
        Emit(new
        {
            @event = "run_end",
            suite,
            passed = results.Sum(result => result.Passed),
            failed = results.Sum(result => result.Failed),
            skipped = results.Sum(result => result.Skipped),
            unimplemented = results.Sum(result => result.Unimplemented),
            duration_ms = (long)(DateTimeOffset.UtcNow - started).TotalMilliseconds,
        });
    }

    private static async Task<GroupResult> RunGroupAsync(string suite, string endpoint, string region, TestGroup group)
    {
        var context = new TestContext(endpoint, region, Environment.GetEnvironmentVariable("OVERCAST_COMPAT_RUN_ID") ?? "local");
        var result = new GroupResult();

        if (group.Setup is not null)
        {
            try
            {
                await group.Setup(context);
            }
            catch (Exception ex)
            {
                var reason = $"setup failed: {ex.Message}";
                foreach (var test in group.Tests)
                {
                    EmitTestResult(suite, group, test.Name, "skip", 0, reason);
                    result.Skipped++;
                }

                await RunTeardownAsync(group, context);
                return result;
            }
        }

        var blocked = new HashSet<string>(StringComparer.Ordinal);
        foreach (var test in group.Tests)
        {
            if (!string.IsNullOrWhiteSpace(test.Skip))
            {
                EmitTestResult(suite, group, test.Name, "skip", 0, test.Skip);
                result.Skipped++;
                blocked.Add(test.Name);
                continue;
            }

            var failedDeps = test.Depends.Where(blocked.Contains).ToList();
            if (failedDeps.Count > 0)
            {
                EmitTestResult(suite, group, test.Name, "skip", 0, $"dependency failed: {string.Join(", ", failedDeps)}");
                result.Skipped++;
                blocked.Add(test.Name);
                continue;
            }

            var started = DateTimeOffset.UtcNow;
            try
            {
                await test.Fn(context);
                EmitTestResult(suite, group, test.Name, "pass", (long)(DateTimeOffset.UtcNow - started).TotalMilliseconds, null);
                result.Passed++;
            }
            catch (Exception ex)
            {
                var status = IsUnimplemented(ex) ? "unimplemented" : "fail";
                EmitTestResult(suite, group, test.Name, status, (long)(DateTimeOffset.UtcNow - started).TotalMilliseconds, ex.Message);
                if (status == "unimplemented")
                {
                    result.Unimplemented++;
                }
                else
                {
                    result.Failed++;
                }
                blocked.Add(test.Name);
            }
        }

        await RunTeardownAsync(group, context);
        return result;
    }

    private static async Task RunTeardownAsync(TestGroup group, TestContext context)
    {
        if (group.Teardown is null)
        {
            return;
        }

        try
        {
            await group.Teardown(context);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[dotnet-sdk] teardown failed for {group.Name}: {ex.Message}");
        }
    }

    public static bool IsUnimplemented(Exception exception)
    {
        var message = exception.ToString();
        return message.Contains("501", StringComparison.OrdinalIgnoreCase)
            || message.Contains("NotImplemented", StringComparison.OrdinalIgnoreCase)
            || message.Contains("UnknownOperationException", StringComparison.OrdinalIgnoreCase)
            || message.Contains("Unknown action", StringComparison.OrdinalIgnoreCase)
            || message.Contains("not implemented", StringComparison.OrdinalIgnoreCase);
    }

    private static void EmitTestResult(string suite, TestGroup group, string test, string status, long durationMs, string? error)
    {
        Emit(new
        {
            @event = "test_result",
            suite,
            service = group.Service,
            group = group.Name,
            test,
            status,
            duration_ms = durationMs,
            error,
        });
    }

    private static readonly object EmitLock = new();

    private static void Emit(object value)
    {
        lock (EmitLock)
        {
            Console.Out.WriteLine(JsonSerializer.Serialize(value, JsonOptions));
            Console.Out.Flush();
        }
    }

    private sealed class GroupResult
    {
        public int Passed { get; set; }
        public int Failed { get; set; }
        public int Skipped { get; set; }
        public int Unimplemented { get; set; }
    }
}

File: compat/suites/java-sdk/src/main/java/io/overcast/compat/harness/Runner.java
package io.overcast.compat.harness;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Runner executes a list of {@link TestGroup}s and emits NDJSON events to
 * {@code stdout}.
 *
 * <p>Events follow the same schema used by the node-js-sdk and go-sdk suites
 * so that the Overcast compatibility dashboard can aggregate results across
 * all language suites.
 *
 * <pre>
 * {"event":"run_start",   "suite":"java-sdk", "started_at":"...", "endpoint":"...", "version":"0"}
 * {"event":"test_result", "suite":"java-sdk", "service":"s3",  "group":"s3-crud",
 *                          "test":"CreateBucket", "status":"pass", "duration_ms":42}
 * {"event":"run_end",     "suite":"java-sdk", "passed":1, "failed":0, "skipped":0,
 *                          "unimplemented":0, "duration_ms":200}
 * </pre>
 */
public final class Runner {

    // Statuses
    private static final String PASS          = "pass";
    private static final String FAIL          = "fail";
    private static final String SKIP          = "skip";
    private static final String UNIMPLEMENTED = "unimplemented";

    private static final ObjectMapper MAPPER = new ObjectMapper();

    /** Emits run_start, executes every group in parallel (bounded by OVERCAST_COMPAT_PARALLEL_SLOTS), then emits run_end. */
    public static void runSuite(String suite, String endpoint, List<TestGroup> groups) {
        long suiteStart = System.currentTimeMillis();

        int totalTests = groups.stream().mapToInt(g -> g.tests().size()).sum();
        emit(new RunStartEvent(suite, Instant.now().toString(), endpoint, "1", totalTests));

        // Limit concurrent group execution to avoid overwhelming the emulator.
        // OVERCAST_COMPAT_PARALLEL_SLOTS is injected by the Go runner; default 8.
        int slots = 8;
        String slotsEnv = System.getenv("OVERCAST_COMPAT_PARALLEL_SLOTS");
        if (slotsEnv != null && !slotsEnv.isEmpty()) {
            try { slots = Math.max(1, Integer.parseInt(slotsEnv)); } catch (NumberFormatException ignored) {}
        }

        AtomicInteger passed = new AtomicInteger();
        AtomicInteger failed = new AtomicInteger();
        AtomicInteger skipped = new AtomicInteger();
        AtomicInteger unimplemented = new AtomicInteger();

        ExecutorService pool = Executors.newFixedThreadPool(slots);
        List<Future<?>> futures = new ArrayList<>(groups.size());
        String region = System.getenv("OVERCAST_DEFAULT_REGION") != null
                ? System.getenv("OVERCAST_DEFAULT_REGION") : "us-east-1";
        String runId = System.getenv("OVERCAST_COMPAT_RUN_ID") != null
                ? System.getenv("OVERCAST_COMPAT_RUN_ID") : "local";

        for (TestGroup group : groups) {
            futures.add(pool.submit(() -> {
                int[] counts = runGroup(suite, endpoint, region, runId, group);
                passed.addAndGet(counts[0]);
                failed.addAndGet(counts[1]);
                skipped.addAndGet(counts[2]);
                unimplemented.addAndGet(counts[3]);
            }));
        }

        for (Future<?> f : futures) {
            try {
                // 5-minute per-group timeout: prevents one hung AWS SDK call
                // from blocking the entire suite indefinitely.
                f.get(5, TimeUnit.MINUTES);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            } catch (ExecutionException | TimeoutException ignored) {}
        }
        pool.shutdown();

        long totalMs = System.currentTimeMillis() - suiteStart;
        emit(new RunEndEvent(suite, passed.get(), failed.get(), skipped.get(), unimplemented.get(), totalMs));
    }

    /** Runs a single group synchronously; returns [passed, failed, skipped, unimplemented]. */
    private static int[] runGroup(String suite, String endpoint, String region, String runId, TestGroup group) {
        TestContext ctx = new TestContext(endpoint, region, runId);
        int passed = 0, failed = 0, skipped = 0, unimplemented = 0;

            // Setup phase
            boolean setupOk = true;
            if (group.setup() != null) {
                try {
                    group.setup().run(ctx);
                } catch (Throwable e) {
                    String reason = "setup failed: " + e.getMessage();
                    for (TestCase tc : group.tests()) {
                        emit(new TestResultEvent(suite, group.service(), group.name(),
                                tc.name(), SKIP, 0, reason));
                        skipped++;
                    }
                    setupOk = false;
                }
            }

            if (!setupOk) {
                runTeardown(group, ctx);
                return new int[]{passed, failed, skipped, unimplemented};
            }

            // Test phase
            Set<String> passedTests = new HashSet<>();
            Set<String> failedOrSkipped = new HashSet<>();

            for (TestCase tc : group.tests()) {
                if (tc.skip() != null && !tc.skip().isEmpty()) {
                    emit(new TestResultEvent(suite, group.service(), group.name(),
                            tc.name(), SKIP, 0, tc.skip()));
                    skipped++;
                    failedOrSkipped.add(tc.name());
                    continue;
                }

                // Cascade-skip: if any dependency failed or was skipped, skip this test.
                List<String> missingDeps = new ArrayList<>();
                for (String dep : tc.depends()) {
                    if (failedOrSkipped.contains(dep)) {
                        missingDeps.add(dep);
                    }
                }
                if (!missingDeps.isEmpty()) {
                    String reason = "dependency failed: " + String.join(", ", missingDeps);
                    emit(new TestResultEvent(suite, group.service(), group.name(),
                            tc.name(), SKIP, 0, reason));
                    skipped++;
                    failedOrSkipped.add(tc.name());
                    continue;
                }

                long start = System.currentTimeMillis();
                try {
                    tc.fn().run(ctx);
                    long ms = System.currentTimeMillis() - start;
                    emit(new TestResultEvent(suite, group.service(), group.name(),
                            tc.name(), PASS, ms, null));
                    passed++;
                    passedTests.add(tc.name());
                } catch (Throwable e) {
                    long ms = System.currentTimeMillis() - start;
                    String msg = e.getMessage() != null ? e.getMessage() : e.getClass().getSimpleName();
                    String status = isUnimplemented(e) ? UNIMPLEMENTED : FAIL;
                    emit(new TestResultEvent(suite, group.service(), group.name(),
                            tc.name(), status, ms, msg));
                    if (status.equals(UNIMPLEMENTED)) {
                        unimplemented++;
                    } else {
                        failed++;
                    }
                    failedOrSkipped.add(tc.name());
                }
            }

            runTeardown(group, ctx);
        return new int[]{passed, failed, skipped, unimplemented};
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private static void runTeardown(TestGroup group, TestContext ctx) {
        if (group.teardown() != null) {
            try {
                group.teardown().run(ctx);
            } catch (Throwable e) {
                System.err.println("[java-sdk] teardown failed for " + group.name() + ": " + e.getMessage());
            }
        }
    }

    /**
     * Returns {@code true} when {@code e} signals a 501 / not-implemented
     * response from the Overcast emulator.
     */
    public static boolean isUnimplemented(Throwable e) {
        if (e == null) return false;
        String msg = e.getMessage();
        if (msg == null) msg = e.getClass().getName();
        return msg.contains("501")
                || msg.contains("NotImplemented")
                || msg.contains("UnknownOperationException")
                || msg.contains("Unknown action")
                || msg.contains("not implemented");
    }

    // Serialises {@code v} as a single NDJSON line to stdout.  Must be
    // synchronised because multiple threads could call emit() concurrently;
    // this ensures lines are never interleaved.
    private static synchronized void emit(Object v) {
        try {
            System.out.println(MAPPER.writeValueAsString(v));
            System.out.flush();
        } catch (Exception e) {
            System.err.println("[java-sdk] failed to serialise event: " + e.getMessage());
        }
    }

    // ── Event record types ────────────────────────────────────────────────────

    record RunStartEvent(
            String event,
            String suite,
            String started_at,
            String endpoint,
            String version,
            int total_tests) {

        RunStartEvent(String suite, String startedAt, String endpoint, String version, int totalTests) {
            this("run_start", suite, startedAt, endpoint, version, totalTests);
        }
    }

    record TestResultEvent(
            String event,
            String suite,
            String service,
            String group,
            String test,
            String status,
            long duration_ms,
            String error) {

        TestResultEvent(String suite, String service, String group, String test,
                        String status, long durationMs, String error) {
            this("test_result", suite, service, group, test, status, durationMs, error);
        }

        // Jackson omits null fields automatically when the mapper is configured
        // with INCLUDE_NON_NULL — but the constructor sets the field, so we use a
        // custom serialiser or just accept null in JSON. The dashboard ignores null.
    }

    record RunEndEvent(
            String event,
            String suite,
            int passed,
            int failed,
            int skipped,
            int unimplemented,
            long duration_ms) {

        RunEndEvent(String suite, int passed, int failed, int skipped, int unimplemented, long ms) {
            this("run_end", suite, passed, failed, skipped, unimplemented, ms);
        }
    }
}
